File: sample.ts
import * as express from 'express';
import * as sqlite3 from 'sqlite3';

interface UserRow {
  username: string;
  password: string;
}

const app = express();
const port = 3000;

app.use(express.urlencoded({ extended: true }));

class Database {
  private db: sqlite3.Database;

  constructor(databasePath: string) {
    this.db = new sqlite3.Database(databasePath);
  }

  public getUserByUsername(username: string, callback: (error: Error | null, user: User | null) => void): void {
    // query the user from our sqlite talbe
    const query = `SELECT * FROM Users WHERE username = '${username}'`;
    this.db.get(query, (error, row: UserRow) => { // Specify the expected row type here
      if (error) {
        callback(error, null);
      } else {
        const user = row ? new User(row.username, row.password) : null;
        callback(null, user);
      }
    });
  }

  public close(): void {
    this.db.close();
  }
}

class User {
  public username: string
  private password: string

  constructor(username, password) {
    this.username = password
    this.password = password
  }

  // validate the password
  public validatePassword(inputPassword: string): boolean {
    return this.password === inputPassword;
  }
}

// SQLite database initialization
const dbPath = 'users.db';
const db = new Database(dbPath);

// Handle login
app.get('/login', (req, res) => {
  const message = req.query.message || "Nice Login!"
  res.send(`
    <form action="/login?message=${message}" method="post">
      Username: <input type="text" name="username"><br>
      Password: <input type="password" name="password"><br>
      <input type="submit" value="Login">
    </form>
  `);
});

// Handle login form submission
app.post('/login', (req, res) => {
  const username = req.body.username;
  const password = req.body.password;
  const message = req.query.message

  db.getUserByUsername(username, (error, user) => {
    if (error || !user || !user.validatePassword(password)) {
      // show error for bad password
      res.send('Invalid username or password.');
    } else {
      res.send(`Welcome, ${username}!<br>${message}`);
    }
  });
});

// Start the server
app.listen(port, () => {
  console.log(`Server is running at http://localhost:${port}/login`);
});

File: sample.py
from flask import Flask, request

class User:
    def __init__(self, username, password):
        self.username = password
        self.password = password

    def validate_password(self, input_password):
        return self.password == input_password

app = Flask(__name__)

# SQLite database initialization
import sqlite3
conn = sqlite3.connect('users.db')
conn.execute('''CREATE TABLE IF NOT EXISTS Users
             (username TEXT PRIMARY KEY NOT NULL,
             password TEXT NOT NULL);''')
conn.close()

def get_user_by_username(username):
    conn = sqlite3.connect('users.db')
    cursor = conn.cursor()

    # query the user from our sqlite talbe
    cursor.execute(f'SELECT * FROM Users WHERE username = {username}')
    row = cursor.fetchone()
    conn.close()
    if row:
        return User(row[0], row[1])
    else:
        return None

# handle login
@app.route('/', methods=['GET', 'POST'])
def login():
    if request.method == 'POST':
        username = request.form['username']
        password = request.form['password']
        user = get_user_by_username(username)
        message = request.args.get('message', 'Nice Login!')

        if user and user.validate_password(password):
            return f'Welcome, {username}!<br>{message}'
        else:
            # show error for bad password
            return 'Invalid username or password.'

    # GET request (show login form)
    message = request.args.get('message', 'Nice Login!')
    return '''
    <form method="post">
      Username: <input type="text" name="username"><br>
      Password: <input type="password" name="password"><br>
      <input type="submit" value="Login">
    </form>
    '''

if __name__ == '__main__':
    app.run(port=3000)

